File: packages/core/android/src/debug/java/expo/modules/ama/a11yChecker.kt
package expo.modules.ama

import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat
import expo.modules.kotlin.AppContext
import java.util.Collections
import kotlin.collections.mutableListOf
import kotlin.synchronized

data class A11yIssue(
        val type: RuleAction,
        val rule: Rule,
        val label: String? = null,
        val reason: String? = null,
        val viewId: Int? = null,
)

enum class RuleAction {
    MUST,
    MUST_NOT,
    SHOULD,
    SHOULD_NOT,
    IGNORE
}

enum class Rule {
    CONTRAST_FAILED,
    CONTRAST_FAILED_AAA,
    FLATLIST_NO_COUNT_IN_SINGULAR_MESSAGE,
    FLATLIST_NO_COUNT_IN_PLURAL_MESSAGE,
    MINIMUM_SIZE,
    NO_ACCESSIBILITY_LABEL,
    NO_ACCESSIBILITY_ROLE,
    NO_FORM_LABEL,
    NO_FORM_ERROR,
    NO_KEYBOARD_TRAP,
    NO_UNDEFINED,
    UPPERCASE_TEXT_NO_ACCESSIBILITY_LABEL,
    NO_UPPERCASE_TEXT,
    BOTTOM_SHEET_CLOSE_ACTION,
    INCOMPATIBLE_ACCESSIBILITY_STATE,
    NO_FORM_LABEL_ENDING_WITH_ASTERISK,
    INCOMPATIBLE_ACCESSIBILITY_ROLE
}

val LOGGER_RULES: Map<Rule, RuleAction> =
        mapOf(
                Rule.CONTRAST_FAILED to RuleAction.MUST,
                Rule.CONTRAST_FAILED_AAA to RuleAction.SHOULD,
                Rule.FLATLIST_NO_COUNT_IN_SINGULAR_MESSAGE to RuleAction.SHOULD,
                Rule.FLATLIST_NO_COUNT_IN_PLURAL_MESSAGE to RuleAction.MUST,
                Rule.MINIMUM_SIZE to RuleAction.MUST,
                Rule.NO_ACCESSIBILITY_LABEL to RuleAction.MUST,
                Rule.NO_ACCESSIBILITY_ROLE to RuleAction.MUST,
                Rule.NO_FORM_LABEL to RuleAction.MUST,
                Rule.NO_FORM_ERROR to RuleAction.MUST,
                Rule.NO_KEYBOARD_TRAP to RuleAction.MUST_NOT,
                Rule.NO_UNDEFINED to RuleAction.MUST_NOT,
                Rule.UPPERCASE_TEXT_NO_ACCESSIBILITY_LABEL to RuleAction.MUST_NOT,
                Rule.NO_UPPERCASE_TEXT to RuleAction.MUST_NOT,
                Rule.BOTTOM_SHEET_CLOSE_ACTION to RuleAction.MUST,
                Rule.INCOMPATIBLE_ACCESSIBILITY_STATE to RuleAction.MUST,
                Rule.NO_FORM_LABEL_ENDING_WITH_ASTERISK to RuleAction.MUST_NOT,
                Rule.INCOMPATIBLE_ACCESSIBILITY_ROLE to RuleAction.MUST_NOT
        )

class A11yChecker(private val appContext: AppContext, private val config: AMAConfig) {
    val activity = appContext.activityProvider?.currentActivity

    private val issues = Collections.synchronizedList(mutableListOf<A11yIssue>())
    private val highlighter = Highlight(appContext)
    private lateinit var rootView: View

    public fun performA11yChecks(rootView: View): List<Map<String, Any?>> {
        Logger.info("performA11yChecks", "Performing a11y checks")

        this.rootView = rootView

        rootView.let { root ->
            val oldIssues = synchronized(issues) { issues.toList() }

            synchronized(issues) { issues.clear() }

            traverseAndCheck(root)

            clearFixedIssues(oldIssues)

            if (issues.isNotEmpty()) {
                Logger.debug("performA11yChecks", issues.toString())

                val issuesPayload =
                        issues.map { issue ->
                            mapOf(
                                    "type" to issue.type,
                                    "rule" to issue.rule,
                                    "label" to (issue.label ?: ""),
                                    "reason" to (issue.reason ?: ""),
                                    "viewId" to issue.viewId
                            )
                        }

                return issuesPayload
            }
        }

        return emptyList()
    }

    public fun clearAllIssues() {
        issues.forEach { issue -> issue.viewId?.let { highlighter.clearHighlight(it) } }
    }

    private fun clearFixedIssues(oldIssues: List<A11yIssue>) {
        val fixed = oldIssues.filter { it !in issues }

        fixed.forEach { issue -> issue.viewId?.let { highlighter.clearHighlight(it) } }
    }

    private fun traverseAndCheck(view: View) {
        val className = view.javaClass.name

        // Ignores the debug overlay
        if (className.startsWith("com.facebook.react.views.debuggingoverlay")) {
            return
        }

        checkView(view)

        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                traverseAndCheck(view.getChildAt(i))
            }
        }
    }

    private fun addIssue(rule: Rule, label: String, reason: String, view: View) {
        val action = getRuleAction(rule)

        if (action == RuleAction.IGNORE) {
            return
        }

        val existingIssue = issues.find { it.rule == rule && it.viewId == view.id }

        if (existingIssue == null) {
            view.id.let { id -> highlighter.highlight(id, config.highlight, action) }

            issues.add(
                    A11yIssue(
                            type = action,
                            rule = rule,
                            label = label,
                            reason = reason,
                            viewId = view.id
                    )
            )
        }
    }

    private fun checkView(view: View) {
        val info = view.createAccessibilityNodeInfo()
        val a11yInfo = AccessibilityNodeInfoCompat.wrap(info)

        if (view.isPressable(a11yInfo)) {
            checkForA11yLabel(view, a11yInfo)
            checkForA11yRole(view, a11yInfo)
            checkForMinimumTargetSize(view)
        }

        // if (view is TextView) {
        //     Logger.debug("checkView", "Check for color contrast")
        //
        //     checkColorContrast(view)
        // }
    }

    private fun checkForA11yLabel(view: View, a11yInfo: AccessibilityNodeInfoCompat) {
        val a11yLabel: String? = a11yInfo.contentDescription?.toString()

        if (a11yLabel.isNullOrEmpty()) {
            Logger.error("checkForA11yLabel", view.toString())
            Logger.error("checkForA11yLabel", view.getTextOrContent())

            addIssue(
                    rule = Rule.NO_ACCESSIBILITY_LABEL,
                    label = view.getTextOrContent(),
                    reason = "",
                    view = view
            )
        }
    }

    private fun checkForA11yRole(view: View, a11yInfo: AccessibilityNodeInfoCompat) {
        val className = a11yInfo.className?.toString() ?: view.javaClass.simpleName
        val roleDescription: String? = a11yInfo.roleDescription?.toString()
        val defaultRole =
                when {
                    className.endsWith(".Button") -> "button"
                    className.endsWith(".CheckBox") -> "checkbox"
                    className.endsWith(".EditText") -> "text field"
                    className.endsWith(".ImageView") && view.isClickable -> "image button"
                    else -> null
                }

        val role = roleDescription ?: defaultRole

        if (role.isNullOrEmpty()) {
            Logger.error("checkForA11yRole", view.getTextOrContent())

            addIssue(
                    rule = Rule.NO_ACCESSIBILITY_ROLE,
                    label = view.getTextOrContent(),
                    reason = "",
                    view = view
            )
        }
    }

    private val density: Float
        get() = activity?.resources?.displayMetrics?.density ?: 1f

    /** dp → px */
    private fun dpToPx(dp: Float): Int = (dp * density + 0.5f).toInt()

    /** px → dp */
    private fun pxToDp(px: Int): Float = px / density

    private fun checkForMinimumTargetSize(view: View) {
        val absBounds = Rect().also { view.createAccessibilityNodeInfo().getBoundsInScreen(it) }

        getHitSlopRect(view)?.let { hitSlop ->
            absBounds.left -= hitSlop.left
            absBounds.top -= hitSlop.top
            absBounds.right += hitSlop.right
            absBounds.bottom += hitSlop.bottom
        }

        val widthPx = absBounds.width()
        val heightPx = absBounds.height()
        val widthDp = widthPx / view.resources.displayMetrics.density
        val heightDp = heightPx / view.resources.displayMetrics.density

        // 4) check vs 48dp
        if (widthDp < 48 || heightDp < 48) {
            addIssue(
                    rule = Rule.MINIMUM_SIZE,
                    label = view.toString(),
                    reason = String.format("%.1f×%.1f dp (< 48 dp)", widthDp, heightDp),
                    view = view
            )
        }
    }

    private fun checkColorContrast(textView: TextView) {
        try {
            // Get text color
            val textColor = textView.currentTextColor

            // Get background color
            val backgroundColor = getBackgroundColor(textView)

            // Calculate contrast ratio
            val contrastRatio = calculateContrastRatio(textColor, backgroundColor)

            // Check against WCAG standards
            val textSize =
                    textView.textSize / textView.context.resources.displayMetrics.scaledDensity
            val isLargeText = textSize >= 18f || (textSize >= 14f && isTextBold(textView))

            val minContrast = if (isLargeText) 3.0 else 4.5 // WCAG AA standards

            if (contrastRatio < minContrast) {
                addIssue(
                        rule = Rule.CONTRAST_FAILED,
                        label = textView.toString(),
                        reason = "Color contrast ratio ${String.format("%.2f", contrastRatio)}",
                        view = textView
                )
            }
        } catch (e: Exception) {
            // Handle cases where color extraction fails
            // Could add logging here if needed
        }
    }

    private fun getBackgroundColor(view: View): Int {
        // Try to get background color from the view itself
        val background = view.background
        if (background is ColorDrawable) {
            return background.color
        }

        // If no background, traverse up the view hierarchy
        var parent = view.parent
        while (parent is View) {
            val parentView = parent as View
            val parentBackground = parentView.background
            if (parentBackground is ColorDrawable) {
                return parentBackground.color
            }
            parent = parentView.parent
        }

        // Default to white if no background found
        return Color.WHITE
    }

    private fun isTextBold(textView: TextView): Boolean {
        val typeface = textView.typeface
        return typeface != null && typeface.isBold
    }

    private fun calculateContrastRatio(color1: Int, color2: Int): Double {
        val luminance1 = calculateLuminance(color1)
        val luminance2 = calculateLuminance(color2)

        val lighter = maxOf(luminance1, luminance2)
        val darker = minOf(luminance1, luminance2)

        return (lighter + 0.05) / (darker + 0.05)
    }

    private fun calculateLuminance(color: Int): Double {
        // Extract RGB components
        val red = Color.red(color) / 255.0
        val green = Color.green(color) / 255.0
        val blue = Color.blue(color) / 255.0

        // Apply gamma correction
        val r = if (red <= 0.03928) red / 12.92 else Math.pow((red + 0.055) / 1.055, 2.4)
        val g = if (green <= 0.03928) green / 12.92 else Math.pow((green + 0.055) / 1.055, 2.4)
        val b = if (blue <= 0.03928) blue / 12.92 else Math.pow((blue + 0.055) / 1.055, 2.4)

        // Calculate relative luminance
        return 0.2126 * r + 0.7152 * g + 0.0722 * b
    }

    private fun getRuleAction(rule: Rule): RuleAction {
        val overrideAction = config.rules[rule.name]

        if (overrideAction != null) {
            return when (overrideAction.uppercase()) {
                "PLEASE_FORGIVE_ME" -> RuleAction.IGNORE
                "MUST" -> RuleAction.MUST
                "SHOULD" -> RuleAction.SHOULD
                "SHOULD_NOT" -> RuleAction.SHOULD_NOT
                "MUST_NOT" -> RuleAction.MUST_NOT
                else -> LOGGER_RULES[rule]!!
            }
        }

        return LOGGER_RULES[rule]!!
    }
}

fun View.isPressable(a11yInfo: AccessibilityNodeInfoCompat): Boolean {
    return this.isClickable && this.isAccessible(a11yInfo)
}

fun View.isAccessible(a11yInfo: AccessibilityNodeInfoCompat): Boolean {
    if (this.importantForAccessibility == View.IMPORTANT_FOR_ACCESSIBILITY_NO ||
                    !a11yInfo.isVisibleToUser
    ) {
        return false
    }

    return true
}

fun View.getTextOrContent(): String {
    if (!this.contentDescription.isNullOrEmpty()) {
        return this.contentDescription.toString()
    }

    if (this is TextView) {
        if (!this.text.isNullOrEmpty()) {
            return this.text.toString()
        }

        if (!this.hint.isNullOrEmpty()) {
            return this.hint.toString()
        }
    }

    val info = createAccessibilityNodeInfo()
    val a11yInfo = AccessibilityNodeInfoCompat.wrap(info)

    return a11yInfo.contentDescription?.toString().orEmpty()
}

fun getHitSlopRect(view: View): Rect? {
    return try {
        val rvClass = Class.forName("com.facebook.react.views.view.ReactViewGroup")

        if (!rvClass.isInstance(view)) {
            Logger.info("getHitSlopRect", "no class found")

            return null
        }

        val getter = rvClass.getMethod("getHitSlopRect")

        @Suppress("UNCHECKED_CAST") val rect = getter.invoke(view) as? Rect

        rect
    } catch (e: ClassNotFoundException) {
        null
    } catch (e: NoSuchMethodException) {
        null
    } catch (e: Exception) {
        null
    }
}

File: packages/core/ios/A11yChecker.swift
import ExpoModulesCore
import UIKit

public struct A11yIssue: Equatable {
    public let type: RuleAction
    public let rule: Rule
    public let label: String?
    public let reason: String?
    public let viewId: Int
    public var sent: Bool

    public static func == (lhs: A11yIssue, rhs: A11yIssue) -> Bool {
        return lhs.viewId == rhs.viewId && lhs.rule == rhs.rule
    }
}

public enum RuleAction: String {
    case must = "MUST"
    case mustNot = "MUST_NOT"
    case should = "SHOULD"
    case shouldNot = "SHOULD_NOT"
    case ignore = "IGNORE"
}

public enum Rule: String {
    case contrastFailed = "CONTRAST_FAILED"
    case contrastFailedAAA = "CONTRAST_FAILED_AAA"
    case flatlistNoCountInSingularMessage = "FLATLIST_NO_COUNT_IN_SINGULAR_MESSAGE"
    case flatlistNoCountInPluralMessage = "FLATLIST_NO_COUNT_IN_PLURAL_MESSAGE"
    case minimumSize = "MINIMUM_SIZE"
    case noAccessibilityLabel = "NO_ACCESSIBILITY_LABEL"
    case noAccessibilityRole = "NO_ACCESSIBILITY_ROLE"
    case noFormLabel = "NO_FORM_LABEL"
    case noFormError = "NO_FORM_ERROR"
    case noKeyboardTrap = "NO_KEYBOARD_TRAP"
    case noUndefined = "NO_UNDEFINED"
    case uppercaseTextNoAccessibilityLabel = "UPPERCASE_TEXT_NO_ACCESSIBILITY_LABEL"
    case noUppercaseText = "NO_UPPERCASE_TEXT"
    case bottomSheetCloseAction = "BOTTOM_SHEET_CLOSE_ACTION"
    case incompatibleAccessibilityState = "INCOMPATIBLE_ACCESSIBILITY_STATE"
    case noFormLabelEndingWithAsterisk = "NO_FORM_LABEL_ENDING_WITH_ASTERISK"
    case incompatibleAccessibilityRole = "INCOMPATIBLE_ACCESSIBILITY_ROLE"
}

private let loggerRules: [Rule: RuleAction] = [
    .contrastFailed: .must,
    .contrastFailedAAA: .should,
    .flatlistNoCountInSingularMessage: .should,
    .flatlistNoCountInPluralMessage: .must,
    .minimumSize: .must,
    .noAccessibilityLabel: .must,
    .noAccessibilityRole: .must,
    .noFormLabel: .must,
    .noFormError: .must,
    .noKeyboardTrap: .mustNot,
    .noUndefined: .mustNot,
    .uppercaseTextNoAccessibilityLabel: .mustNot,
    .noUppercaseText: .mustNot,
    .bottomSheetCloseAction: .must,
    .incompatibleAccessibilityState: .must,
    .noFormLabelEndingWithAsterisk: .mustNot,
    .incompatibleAccessibilityRole: .mustNot,
]

public class A11yChecker {
    private let appContext: AppContext
    private let config: AMAConfig
    private var issues = [A11yIssue]()
    private let highlighter: Highlight
    private var newIssues = false

    public init(appContext: AppContext, config: AMAConfig) {
        self.appContext = appContext
        self.config = config
        self.highlighter = Highlight()
    }

    public func performA11yChecks(on rootView: UIView?) -> [String: Any] {
        guard let root = rootView else { return [:] }

        let oldIssues = issues.map { $0 }

        issues.removeAll()

        traverseAndCheck(view: root)
        let shouldSendEvent = clearFixedIssues(oldIssues)

        let newIssuesToSend = issues.filter { !$0.sent }

        if !newIssuesToSend.isEmpty {
            for newIssue in newIssuesToSend {
                if let index = issues.firstIndex(of: newIssue) {
                    issues[index].sent = true
                }
            }

            return [
                "issues": newIssuesToSend.map { issue in
                    [
                        "type": issue.type.rawValue,
                        "rule": issue.rule.rawValue,
                        "reason": issue.reason ?? "",
                        "label": issue.label ?? "",
                        "viewId": issue.viewId,
                    ]
                },
                "sendEvent": true,
            ]
        }

        return [
            "issues": [],
            "sendEvent": shouldSendEvent,
        ]
    }

    public func clearAllHighlights() {
        issues.forEach { issue in
            highlighter.clearHighlight(viewId: issue.viewId)
        }

        issues.removeAll()
    }

    private func clearFixedIssues(_ oldIssues: [A11yIssue]) -> Bool {
        let fixed = oldIssues.filter {
            old in !issues.contains(old)
        }

        fixed.forEach { issue in
            highlighter.clearHighlight(viewId: issue.viewId)
        }

        for idx in issues.indices {
            if oldIssues.contains(issues[idx]) {
                issues[idx].sent = true
            }
        }

        return !fixed.isEmpty
    }

    private func traverseAndCheck(view: UIView) {
        checkView(view)

        for subview in view.subviews {
            traverseAndCheck(view: subview)
        }
    }

    private func checkView(_ view: UIView) {
        if view.isPressable {
            checkForA11yLabel(view)
            checkForA11yRole(view)
            checkForMinimumTargetArea(view)
            checkColorContrast(on: view)
        }

        // if let label = view as? UILabel {
        //   Logger.debug("checkView", "Check for color contrast")
        //
        //   checkColorContrast(label)
        // }
    }

    private func checkForA11yLabel(_ view: UIView) {
        if view.accessibilityLabel?.isEmpty ?? true {
            addIssue(
                rule: .noAccessibilityLabel,
                label: view.getTextOrContent(),
                reason: "",
                view: view
            )
        }
    }

    private func checkForA11yRole(_ view: UIView) {
        let defaultRole: String? = {
            switch view {
            case is UIButton:
                return "button"
            case is UISwitch:
                return "switch"
            case is UITextField:
                return "text field"
            case let iv as UIImageView where iv.isUserInteractionEnabled:
                return "image button"
            default:
                return nil
            }
        }()

        let traits = view.accessibilityTraits
        let hasRoleTrait =
            traits.contains(.button)
            || traits.contains(.link)
            || traits.contains(.searchField)
            || traits.contains(.image)

        if !(hasRoleTrait || defaultRole != nil) {
            let label = view.getTextOrContent()
            let traitNames = view.accessibilityTraits.names

            addIssue(
                rule: .noAccessibilityRole,
                label: label,
                reason: "",
                view: view
            )
        }
    }

    private func checkForMinimumTargetArea(_ view: UIView) {
        let baseSize: CGRect = view.frame
        let insets = view.getHitSlopRect()
        let pressableWidth = baseSize.width - insets.left - insets.right
        let pressableHeight = baseSize.height - insets.top - insets.bottom

        if view.isPressable && (pressableWidth < 44 || pressableHeight < 44) {
            addIssue(
                rule: .minimumSize,
                label: view.getTextOrContent(),
                reason:
                    "\(Int(pressableWidth))x\(Int(pressableHeight))",
                view: view
            )
        }
    }

    private func checkColorContrast(on view: UIView) {
        guard
            let fg = view.contentColor
        else {
            return
        }
        let bg = getBackgroundColor(for: view)
        let contrastRatio = calculateContrastRatio(color1: fg, color2: bg)
        var minContrast = 3.0

        if let font = view.contentFont {
            let textSize = font.pointSize
            let isBold = font.fontDescriptor.symbolicTraits.contains(.traitBold)
            let isLargeText = textSize >= 18 || (textSize >= 14 && isBold)

            minContrast = isLargeText ? 3.0 : 4.5
        }

        if contrastRatio < minContrast {
            addIssue(
                rule: .contrastFailed,
                label: view.getTextOrContent(),
                reason: String(
                    format:
                        "Color contrast ratio %.2f is below minimum %.1f (WCAG AA).\nForeground %@ (%@) on background color %@ (%@)",
                    contrastRatio, minContrast,
                    fg.hexString, fg.rgbaString,
                    bg.hexString, bg.rgbaString
                ),
                view: view
            )
        }
    }

    private func getBackgroundColor(for view: UIView) -> UIColor {
        if let bg = view.backgroundColor { return bg }
        var parent = view.superview
        while let p = parent {
            if let bg = p.backgroundColor { return bg }
            parent = p.superview
        }
        return .white
    }

    private func calculateContrastRatio(color1: UIColor, color2: UIColor) -> Double {
        let l1 = calculateLuminance(color1)
        let l2 = calculateLuminance(color2)
        let lighter = max(l1, l2)
        let darker = min(l1, l2)
        return (lighter + 0.05) / (darker + 0.05)
    }

    private func calculateLuminance(_ color: UIColor) -> Double {
        guard let comps = color.cgColor.components, comps.count >= 3 else { return 1.0 }
        let red = Double(comps[0])
        let green = Double(comps[1])
        let blue = Double(comps[2])
        func adjust(_ c: Double) -> Double {
            return c <= 0.03928 ? c / 12.92 : pow((c + 0.055) / 1.055, 2.4)
        }
        let r = adjust(red)
        let g = adjust(green)
        let b = adjust(blue)
        return 0.2126 * r + 0.7152 * g + 0.0722 * b
    }

    private func getRuleAction(_ rule: Rule) -> RuleAction {
        if let override = config.rules[rule.rawValue]?.uppercased() {
            switch override {
            case "PLEASE_FORGIVE_ME": return .ignore
            case "MUST": return .must
            case "SHOULD": return .should
            case "SHOULD_NOT": return .shouldNot
            case "MUST_NOT": return .mustNot
            default: return loggerRules[rule]!
            }
        }
        return loggerRules[rule]!
    }

    private func addIssue(rule: Rule, label: String, reason: String, view: UIView) {
        let action = getRuleAction(rule)
        let viewId = view.tag
        guard action != .ignore, viewId > 0 else { return }

        if !issues.contains(where: { $0.rule == rule && $0.viewId == viewId }) {
            self.newIssues = true

            issues.append(
                A11yIssue(
                    type: action,
                    rule: rule,
                    label: label,
                    reason: reason,
                    viewId: viewId,
                    sent: false
                ))
        }
    }

    public func clearHighlight(view: UIView) {
        let viewId = view.tag

        highlighter.clearHighlight(viewId: viewId)
    }

    public func highlight(view: UIView) {
        let viewId = view.tag
        if let issue = issues.first(where: { $0.viewId == viewId }) {
            let rule = issue.rule
            let action = getRuleAction(rule)

            highlighter.highlight(view: view, mode: config.highlight, action: action)
        }
    }
}

extension UIView {
    fileprivate func getTextOrContent() -> String {
        if let label = accessibilityLabel, !label.isEmpty {
            return label
        }
        if let lbl = self as? UILabel, let t = lbl.text, !t.isEmpty {
            return t
        }
        if let tf = self as? UITextField, let ph = tf.placeholder, !ph.isEmpty {
            return ph
        }
        return ""
    }

    var contentColor: UIColor? {
        let className = String(describing: type(of: self))

        if className.contains("RCTParagraphComponentView"),
            let anyObj = self as AnyObject?,
            let attrText = anyObj.value(forKey: "attributedText") as? NSAttributedString,
            attrText.length > 0,
            let fgColor = attrText.attribute(.foregroundColor, at: 0, effectiveRange: nil)
                as? UIColor
        {
            return fgColor
        }

        if let label = self as? UILabel {
            return label.textColor
        }
        if let button = self as? UIButton {
            if let tc = button.titleColor(for: .normal) { return tc }
            return button.tintColor
        }
        if let iv = self as? UIImageView {
            return iv.tintColor
        }

        for sub in subviews {
            if let cc = sub.contentColor {
                return cc
            }
        }

        return self.tintColor
    }

    var contentFont: UIFont? {
        if let label = self as? UILabel {
            return label.font
        }
        if let button = self as? UIButton,
            let f = button.titleLabel?.font
        {
            return f
        }
        // Recurse into children
        for sub in subviews {
            if let f = sub.contentFont {
                return f
            }
        }
        return nil
    }

    fileprivate func getHitSlopRect() -> UIEdgeInsets {
        let selector = NSSelectorFromString("hitTestEdgeInsets")

        guard self.responds(to: selector) else {
            return .zero
        }

        let anyObj = self as AnyObject
        guard
            let edgeValue = anyObj.value(forKey: "hitTestEdgeInsets") as? UIEdgeInsets
        else {
            return .zero
        }

        return edgeValue
    }

    /**
     * Returns true if the user can tap on the element
     */
    var isPressable: Bool {
        return isUserInteractionEnabled && isAccessibilityElement
    }
}

extension UIAccessibilityTraits {
    /// Returns the names of all traits present in this bitmask.
    fileprivate var names: [String] {
        let all: [(UIAccessibilityTraits, String)] = [
            (.button, "button"),
            (.link, "link"),
            (.header, "header"),
            (.searchField, "searchField"),
            (.image, "image"),
            (.selected, "selected"),
            (.playsSound, "playsSound"),
            (.keyboardKey, "keyboardKey"),
            (.staticText, "staticText"),
            (.summaryElement, "summaryElement"),
            (.notEnabled, "notEnabled"),
            (.updatesFrequently, "updatesFrequently"),
            (.startsMediaSession, "startsMediaSession"),
            (.adjustable, "adjustable"),
            (.allowsDirectInteraction, "allowsDirectInteraction"),
            (.causesPageTurn, "causesPageTurn"),
        ]
        return all.compactMap { (trait, name) in
            (self.rawValue & trait.rawValue) != 0 ? name : nil
        }
    }
}

extension UIColor {
    var rgbaComponents: (r: CGFloat, g: CGFloat, b: CGFloat, a: CGFloat)? {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0
        guard self.getRed(&r, green: &g, blue: &b, alpha: &a) else { return nil }
        return (r, g, b, a)
    }

    var rgbaString: String {
        guard let c = rgbaComponents else { return self.description }
        let R = Int(c.r * 255)
        let G = Int(c.g * 255)
        let B = Int(c.b * 255)
        return String(format: "rgba(%d,%d,%d,%.2f)", R, G, B, c.a)
    }

    var hexString: String {
        guard let c = rgbaComponents else { return self.description }
        let R = Int(c.r * 255)
        let G = Int(c.g * 255)
        let B = Int(c.b * 255)
        return String(format: "#%02X%02X%02X", R, G, B)
    }
}
